Update ESACCI-OZONE CMORizer(formatter and downloader) for REF

diegokam opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

    While the checklist is intended to be filled in by the technical and scientific
    reviewers, it is the responsibility of the author of the pull request to make
    sure all items on it are properly implemented.
-->

## Description
This PR updates the` ESACCI-OZONE` cmorizer (NCL version to python) for REF, and it will produce 2 aggregated output files for  `toz` (1995-2023) and `o3 ` (1984-2022) variables defined in the CMOR Table [AERmon](https://github.com/ESMValGroup/ESMValCore/blob/main/esmvalcore/cmor/tables/cmip6/Tables/CMIP6_AERmon.json). 
<!--
    Please describe your changes here, especially focusing on why this pull request makes
    ESMValTool better and what problem it solves.

    Before you start, please read our contribution guidelines: https://docs.esmvaltool.org/en/latest/community/

    Please fill in the GitHub issue that is closed by this pull request, e.g. Closes #1903
-->

* * *

## Before you get started

<!--
    Please discuss your idea with the development team before getting started,
    to avoid disappointment or unnecessary work later. The way to do this is
    to open a new issue on GitHub.
-->

- [ ] [☝ Create an issue](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#contributing-code-and-documentation) to discuss what you are going to do

## Checklist

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🛠][1] This pull request has a [descriptive title](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-title)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#code-quality)
- [x] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#documentation) is available
- [x] [🛠][1] [Tests](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#tests) run successfully
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#list-of-authors) is up to date
- [x] [🛠][1] Any changed dependencies have been [added or removed correctly](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#dependencies)
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-checks) were successful

### [New or updated data reformatting script](https://docs.esmvaltool.org/en/latest/develop/dataset.html)

- [x] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/dataset.html#dataset-documentation) is available
- [x] [🛠][1] The dataset has been [added to the CMOR check recipe](https://docs.esmvaltool.org/en/latest/community/dataset.html#testing)
- [x] [🛠][1] The dataset has been added to the shared [data pools](https://docs.esmvaltool.org/en/latest/community/dataset.html#cmorized-data) of DKRZ and Jasmin by the @ESMValGroup/OBS-maintainers team
- [x] [🧪][2] Numbers and units of the data look [physically meaningful](https://docs.esmvaltool.org/en/latest/community/dataset.html#scientific-sanity-check)

***

To help with the number of pull requests:

-  🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValTool/pulls) in this repository

<!--
If you need help with any of the items on the checklists above, please do not hesitate to ask by commenting in the issue or pull request.
-->

schlunma reviewed (changes requested):
Thanks Diego!

I have a couple of comments on the code. In addition, please also call the function `fix_coords` (if possible) on the data to make sure that all coordinate metadata is set properly.

diegokam commented on esmvaltool/cmorizers/data/cmor_config/ESACCI-OZONE.yml:
```diff
@@ -0,0 +1,26 @@
+  tier: 2
+  modeling_realm: sat
+  project_id: OBS
+  source: ""
```
Not sure if this one` ftp://anon-ftp.ceda.ac.uk/neodc/esacci/ozone/data/` indicate in the `/esmvaltool/cmorizers/data/datasets.yml` for the downloader is still valid as I tested the new cmorizer directly on the dataset available on Levante. Maybe @axel-lauer can confirm?

axel-lauer replied:
The versions I provided are from the Copernicus Climate Data Store (CDS) as instructed by the ESA CCI team. If exactly the same versions are available from CEDA, then we could use this as download link. If not, I can provide the download instructions for CDS.

schlunma replied:
This is not resolved. If the link differs for the two variables, you need to put this under the corresponding variables.

diegokam replied:
@schlunma Is the documentation fine now? It also passes the doc build tests that yesterday were failing. As the download links are the same (i.e. CDS website) for both variables I just added the variable names close to the dataset name and kept the structure suggested by Axel.

diegokam commented on esmvaltool/cmorizers/data/formatters/datasets/esacci_ozone_old.ncl:
Not sure if we need to keep as we did for other cmorizers like ESA CCI Landcover

schlunma replied:
In my opinion this makes only sense if it can be used with the CMORizer interface (`esmvaltool data format ...`). I fear it won't...If it will, then the dataset's name will also be `ESACCI-OZONE-v0002`, which will be different from the one used in the recipes. Unfortunately we still haven't found a way how to [deal with different dataset versions](https://github.com/ESMValGroup/ESMValTool/issues/3452) properly. Is it really that useful to have this old file? Did we really keep that for the other CMORizers as well?

axel-lauer replied:
Technically, this would be a nice case for the "archive" folder that we proposed in the last science lead development team meeting (https://github.com/ESMValGroup/Community/blob/main/Science%20Lead/Minutes/20250127.md).

diegokam replied:
Should I create a dedicated folder as indicated in the meeting notes and move the old NCL cmorizer there? As from the notes it seems there is not a final decision yet, right?

schlunma replied:
For the other CMORizer PRs (https://github.com/ESMValGroup/ESMValTool/pull/3676, https://github.com/ESMValGroup/ESMValTool/pull/3727), we just removed the old NCL code. I think it would be a good idea to do that here, too.

diegokam commented on esmvaltool/cmorizers/data/cmor_config/ESACCI-OZONE.yml:
```diff
@@ -0,0 +1,26 @@
+  modeling_realm: sat
+  project_id: OBS
+  source: ""
+  reference: "esacci-ozone" 
```
I found this [one ](https://amt.copernicus.org/preprints/amt-2024-196/) but not sure if it is the main reference indeed

schlunma replied:
@axel-lauer what do you think?

axel-lauer replied:
Not sure, but shouldn't `project_id` rather be `OBS6?` An entry for `source` is missing and should point to the Copernicus Climate Data Store, e.g.  https://cds.climate.copernicus.eu/datasets/satellite-ozone-v1

Regarding a data reference, this is what CDS recommends:

Copernicus Climate Change Service, Climate Data Store, (2020): Ozone monthly gridded data from 1970 to present derived from satellite observations. Copernicus Climate Change Service (C3S) Climate Data Store (CDS). DOI: [10.24381/cds.4ebfe4eb](https://doi.org/10.24381/cds.4ebfe4eb) (Accessed on DD-MMM-YYYY)

diegokam replied:
Yes, the` project_id` is `OBS6` in my latest [push ](https://github.com/ESMValGroup/ESMValTool/pull/3899/files#diff-038c8d754ea34765e2c3a5e7159170c8ec4ceffc604b8b23899c45abe1c7c5ec) indeed, maybe this is still referring to the previous one.  I am going to update the data reference, thanks!

schlunma commented on esmvaltool/cmorizers/data/formatters/datasets/esacci_ozone.py:
```diff
@@ -0,0 +1,147 @@
+    if cmor_info is None:
+        raise ValueError(f"CMOR info for {short_name} in MIP {mip} not found!")
+
+    year = year
+    month = month
```
```suggestion
```

schlunma commented on esmvaltool/cmorizers/data/formatters/datasets/esacci_ozone.py:
```diff
@@ -0,0 +1,147 @@
+    day = 15  # Mid-month
+    time_units = Unit("days since 1950-01-01")
+    time_points = time_units.date2num(datetime(year, month, day))
+    print(f"Filename: {filename}, Extracted Year: {year}, Month: {month}")
```
Either use logger or remove

schlunma commented on esmvaltool/cmorizers/data/formatters/datasets/esacci_ozone.py:
```diff
@@ -0,0 +1,147 @@
+        for year in range(start_date.year, end_date.year + 1):
+            for month in range(1, 13):
+                date_str = f"{year}{month:02}"  # YYYYMM format
+
+                # Create glob pattern, handling prefixes:
+                fname_pattern = f"*{date_str}-{var['filename']}"
+                fname = glob.glob(os.path.join(in_dir, fname_pattern))
+
+                if not fname:
+                    fname = glob.glob(os.path.join(in_dir, fname_pattern))
+                    if not fname:
+                        raise ValueError(
+                            "No file found for %s in %s-%02d", var_name, year,
+                            month
+                        )
+
+                filename = fname[0]  # Take the first matching file (if any)
```
```suggestion
                filename = Path(in_dir) / f"{date_str}-{var['filename']}"
                if not filename.is_file():
                    raise ValueError(
                        "No file found for %s in %s-%02d",
                        var_name,
                        year,
                        month,
                    )
```

I don't think you need `glob` here, the filename should be exactly determined by the date string and the file name given in the config file. You need to `from pathlib import Path` for this.

schlunma reviewed (commented):
Thanks Diego for addressing my comments!

Just minor ones remaining. Please make sure to add the appropriate download instructions and source (I am sure @axel-lauer will point you to the right location).

I haven't tested this though.

diegokam commented:
Following clarification with @schlunma, I had to switch to OBS6 project as by running the check_obs on the latest output files the checks were failing because during the CMORization I was using the project OBS that is referring to the CMIP5 tables and not to the CMIP6 ones. Furthermore I had also to use the `fix_alt16_metadata` function from `esmvalcore.cmor._fixes.native_datasets`. 
The output filenames will have now this naming:
`OBS6_ESACCI-OZONE_sat_L3_AERmon_o3_YYYYMM-YYYYMM.nc`
`OBS6_ESACCI-OZONE_sat_L3_AERmon_toz_YYYYMM-YYYYMM.nc`  
The new output files are passing the check_obs tests now.

schlunma commented on esmvaltool/cmorizers/data/formatters/datasets/esacci_ozone.py:
```diff
@@ -0,0 +1,167 @@
+    20250107
+
+Download and processing instructions
+    Download the data from:
```
```suggestion
```

schlunma commented on esmvaltool/cmorizers/data/formatters/datasets/esacci_ozone.py:
```diff
@@ -0,0 +1,167 @@
+Download and processing instructions
+    Download the data from:
+
+    GTO-ECV total column
```
Mention which variable that is (I guess `toz`), same for the one below.

schlunma reviewed (commented):
Code looks good now. Will approve once the documentation is updated properly. Please also update the [`datasets.yml`](https://github.com/ESMValGroup/ESMValTool/blob/main/esmvaltool/cmorizers/data/datasets.yml#L522) file.

schlunma reviewed (approved):
Code looks good! Didn't run the code nor looked at the output from a scientific point of view, though.

Also the link to the data does not work for me: https://cds.climate.copernicus.eu/datasets/satellite-ozone-v1

axel-lauer commented on esmvaltool/cmorizers/data/formatters/datasets/esacci_ozone.py:
```diff
@@ -0,0 +1,166 @@
+        cube = mixing_ratio / (1.0 + mixing_ratio)
+        cube.units = 'mol mol-1'
+
+    elif short_name == 'toz':  # Total ozone column (DU or m)
+        cube = cube * 2241.399
+        cube.units = 'DU'
+        cube = cube * 2.24115e-5
+        cube.units = 'm'
```
I think the unit conversion is not correct. The total ozone column should be converted from mol m-2 to m (CMOR unit for toz):

```suggestion
    elif short_name == 'toz':  # Total ozone column (m)
        # Convert from mol m-2 to m
        # -------------------------
        # 1e-5 m (gas @ T = 273 K and p = 101325 Pa) ~ 2.69e20 molecules m-2
        # (see https://en.wikipedia.org/wiki/Dobson_unit)
        # 1 m (O3 @ standard conditions) = 1e5 * 2.69e20 molecules m-2
        #                              = 2.69e25 moelcules m-2 / N_A mol m-2
        #                              = 2.69e25 molecules m-2 /
        #                                    (6.022e23 molecules mol-1) mol m-2
        #                              = 44.67 mol m-2
        cube = cube / 44.67
        cube.units = 'm'
```

axel-lauer reviewed (changes requested):
Here are a few first comments:
- I think the unit conversion needs some changes (see code suggestions).
- The default time range in `cmorization` is 1984/1/1 to 2023/12/31. This needs to be changed depending on the dataset that is processed (formatter does not run otherwise):
   - GTO-ECV covers the period 1995/7 to 2023/04
   - SAGE-CCI-OMPS covers the period 1984/10 to 2022/12
- The old downloader is still the one for the old dataset (used with the NCL formatter). As a minimum, the downloader needs to be removed but I would prefer to add a downloader to retrieve the current data from the CDS. Examples for retrieving data from CDS are, e.g. cds_satellite_soil_moisture.py or cds_xch4.py. I believe they can be used as a template with some small modifications, see #3750. Maybe @bettina-gier can advice what is needed to download data from the (new) CDS.

diegokam commented:
Thank you for your review @axel-lauer. 
I do confirm that I have addressed all your comments regarding the formatter and it should working properly now. 
Regarding the downloader I also pushed a working version to download the dataset from CDS but not sure if this could be a proper solution as the new APIs requires this `.cdsapirc` file, as mentioned by @bettina-gier indeed, but it needs to be saved in my home directory and it contains the key assigned to my account registration in the CDS. 
This downloader at the moment is creating two zip files containing all the files over the timepriods for the `o3 `and `toz ` variables and try to use a function in the esmavaltool to automatically extract them in the `OBS `folder.

bettina-gier commented:
Regarding CDS

- I checked their forum yesterday and there was a post in November about a solution for the different urls coming "soon".. who knows when zoom is. In [this post](https://forum.ecmwf.int/t/cds-ads-and-ewds-down-until-further-notice/8015/39) they say you can pass the url when calling the client, I did not check if it's possible to only pass the url and not the key.
- The .cdsapirc file in the home folder was always needed for the CDS downloaders, so that is fine, the only problem is not wanting to change the url in someones home folder file through ESMValTool.
- I also saw they split the format key from e.g. `format: 'netcdf.zip'` to `data_format: 'netcdf'` and `download_format: 'unarchived'` not sure if this applies here, didn't look into the code. In case you didn't want to download zips you have to unpack.

diegokam commented:
Hi @axel-lauer and @bettina-gier, thank you for your inouts, I do confirm that also the downloader should be complete now. 
It doesn't need the `.cdsapirc` file anymore saved in the user HOME but the key associated with the ECMWF account requirest to download the data needs to be added at `line #18`. 
This version download the files for `o3 `and `toz `as two .gz files in the RAWOBS dedicated directory and extract all the files in the same directory. I couldn't use the `unpack_files_in_folder` function already present in the ESMValTool, to do so as I was keep getting an error during the files extraction that might be related to the new . 
I have worked also on another working version for the downloader that is downloading all the files one by one but it is quite slow compared to this one that I have just pushed. Looking forward to your review then! Thanks!

bettina-gier commented:
I just checked with your code, the key is not necessary to be included in the client call line 41, you can give the url alone and then it grabs the key from your `.cdsapirc` file in the home directory. I would prefer that option than having each user manually add their key in the downloader scripts, as we have several cmorizer for CDS and ADS data.

schlunma commented on esmvaltool/cmorizers/data/downloaders/datasets/esacci_ozone_v0002.py:
I don't think we need this (we deleted the old CMORizer).

axel-lauer reviewed (approved):
With the last few small changes (https://github.com/ESMValGroup/ESMValTool/commit/ea95d52f27a91e0102dfd53cc8214214fe976be5), the output looks all fine now. Thanks @diegokam for the hard work!

schlunma reviewed (changes requested):
Could you please try to fix the [Codacy issues](https://app.codacy.com/gh/ESMValGroup/ESMValTool/pull-requests/3899/issues) that are easy to fix? There are quite a lot, i.e.,

- Variable name "r" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern ('[a-z_][a-z0-9_]{2,30}$' pattern) (invalid-name)
- Exception arguments suggest string formatting might be intended (raising-format-tuple)
- Unused Path imported from pathlib (unused-import)
- No blank lines allowed after function docstring (found 2) (D202)
- Variable name "f" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern ('[a-z_][a-z0-9_]{2,30}$' pattern) (invalid-name)
- Exception arguments suggest string formatting might be intended
- Variable name "f" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern ('[a-z_][a-z0-9_]{2,30}$' pattern)
- standard import "pathlib.Path" should be placed before third party imports "iris", "iris.util", "cf_units.Unit" (simply run `isort`)

diegokam commented:
> Could you please try to fix the [Codacy issues](https://app.codacy.com/gh/ESMValGroup/ESMValTool/pull-requests/3899/issues) that are easy to fix? There are quite a lot, i.e.,
> 
> * Variable name "r" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern ('[a-z_][a-z0-9_]{2,30}$' pattern) (invalid-name)
> * Exception arguments suggest string formatting might be intended (raising-format-tuple)
> * Unused Path imported from pathlib (unused-import)
> * No blank lines allowed after function docstring (found 2) (D202)
> * Variable name "f" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern ('[a-z_][a-z0-9_]{2,30}$' pattern) (invalid-name)
> * Exception arguments suggest string formatting might be intended
> * Variable name "f" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern ('[a-z_][a-z0-9_]{2,30}$' pattern)
> * standard import "pathlib.Path" should be placed before third party imports "iris", "iris.util", "cf_units.Unit" (simply run `isort`)

Thanks @schlunma, I have just fixed them as they were not identified by the `flake8` indeed. The remaining codacy issues related to the `cmorization `and `extract_variable ` functions cannot be  fixed easily as for other cmorizers.

schlunma reviewed (approved):
Thanks Diego, looks good now 🎉
